### DISCUSSION TITLE
Issue compiling with Nuitka

### DISCUSSION THREAD

**Discussion by RedTTG** [pymupdf/PyMuPDF#4127] [Ideas]
### Description of the bug

`pymupdf.mupdf` fails to compile with nuitka due to the massive c file it generates.
I have mentioned this issue to nuitka but maybe you can take a look at it too
https://github.com/Nuitka/Nuitka/issues/3243

### How to reproduce the bug

install python 3.9 (tested in other versions and OSes too)
`pip install nuitka pymupdf`
have a `test.pdf` and a simple code as such in a file for example named `test_pymupdf.py`: 
```py
import pymupdf

pdf = pymupdf.open('test.pdf', filetype='pdf')

pdf_page = pdf[0]

pdf_page.get_pixmap().writePNG('test.png')
```

try to compile it with nuitka:
`nuitka --mode=app --script-name=test_pymupdf.py`
PASS 1 will take an absurdly long time (hour or so)
Then the compilation of the massive c file for mupdf will usually overload github action memory, on a beefy machine it might compile but probably after a few hours

### PyMuPDF version

1.25.0

### Operating system

Windows

### Python version

3.9

**Comment by JorjMcKie** [comment]
We are unable to accept this issue as a PyMuPDF problem. It obviously cannot be resolved by whatever is in PyMuPDF's reach.
I am going to move your post to the "Discussions" tab to open it for ideas / contributions other users may have.

**Reply by wenkenwan** [reply]
The old version can be compiled normally.
PyMuPDF v1.22.0

**Reply by RedTTG** [reply]
> The old version can be compiled normally.
> PyMuPDF v1.22.0

https://github.com/JustRedTTG/test-pymupdf/actions/runs/12259960982/job/34203400176

It has issues installing on base windows action container, are there extra steps to install this version?

**Reply by wenkenwan** [reply]
I have not encountered any difficulties, everything has been very smooth. You can try to delete the original and then install the old version.

**Reply by RedTTG** [reply]
> I have not encountered any difficulties, everything has been very smooth. You can try to delete the original and then install the old version.

Read carefully. 
This is about a github action container, I even gave you a link.
Take a second to take a look at it!

**Comment by JorjMcKie** [comment]
Among PyMuPDF's Python sources, there exist 2 (very) large files: `__init__.py`, 23 KLOC, and `mupdf.py`, 60 KLOC.

To compile these to C modules simply takes a long time - how could that ever be changed?
From myself using Nuitka, I remember that there exist options to exclude selected files from compilation.
PyMuPDF's own application code is not part of mupdf.py.
That file is generated automatically by SWIG and wraps the MuPDF C-library. As such, it wouldn't profit from being compiled to C itself. So I suggest to exclude it from Nuitka's compilation.

PyMuPDF's other Python files however should be compiled via Nuitka. That probably still will take long, but the obvious solution is to do it once only (making a `.pyd` module from it) - and not over and over again for each of your apps.

**Reply by RedTTG** [reply]
Yes you can exclude pymypdf.mupdf module but it completely breaks pymupdf, additionally this large compilation breaks GitHub actions. And I do agree since this is an already generated file it shouldn't be then generated again into c to compile it, but I'm not sure how you would go about fixing this. I have a report if you can get the action to work send a PR: https://github.com/JustRedTTG/test-pymupdf

**Comment by zeldridge** [comment]
Has anyone got a solution for this? Having issues with compiling Nuitka with PyMuPDF

**Reply by RedTTG** [reply]
One solution is to close everything on a beefy machine and just compile it it in a few HOURS. The other solution is to just wait for nuitka's promised overhead improvements update to release. But it's still unknown when it'll be ready and if it'll fix this issue.

**Comment by Kushim-Jiang** [comment]
I use v1.18.19 and it works fine.

**Reply by RedTTG** [reply]
Example?

**Comment by jamie-lemon** [comment]
If required please continue this discussion on our forum at https://forum.pymupdf.com , we are trying to move all discussions there!
